cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java

```diff
@@ -143,69 +144,87 @@ private static String httpServletRequestToString(HttpServletRequest request) {
   }
 
   /**
-   * Create the UserTaskInfo reference if it doesn't exist.
+   * Create the UserTask reference for sync or async task if it doesn't exist.
    *
    * This method creates references {@link UserTaskInfo} and maps it to {@link HttpSession} from httpServletRequest. The
    * {@link HttpSession} is also used to fetch {@link OperationFuture} for the in-progress/completed UserTask. If the
    * UserTaskID is passed in the httpServletRequest header then that takes precedence over {@link HttpSession} to fetch
-   * the {@link OperationFuture} for the UserTask.
+   * the {@link OperationFuture} for the UserTask. Sync task is not saved in Session map.
    *
-   * @param httpServletRequest the HttpServletRequest to create the UserTaskInfo reference.
+   * @param httpServletRequest the HttpServletRequest to create the UserTaskInfo reference. Sync request should not contain
+   *                           UUID in the header.
    * @param httpServletResponse the HttpServletResponse that contains the UserTaskId in the HttpServletResponse header.
-   * @param function A function that takes a UUID and returns {@link OperationFuture}.
+   * @param function A function that takes a UUID and returns {@link OperationFuture}. For sync task, the function always
+   *                 returns a completed Future.
    * @param step The index of the step that has to be added or fetched.
+   * @param asyncComplete Indicate whether the task is sync or async.
    * @return The list of {@link OperationFuture} for the linked UserTask.
    */
-  public List<OperationFuture> getOrCreateUserTask(HttpServletRequest httpServletRequest,
-                                                   HttpServletResponse httpServletResponse,
-                                                   Function<String, OperationFuture> function,
-                                                   int step) {
+  public List<OperationFuture> getOrCreateUserTask(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
+      Function<String, OperationFuture> function, int step, boolean asyncComplete) {
     UUID userTaskId = getUserTaskId(httpServletRequest);
-    List<OperationFuture> operationFutures = getFuturesByUserTaskId(userTaskId, httpServletRequest);
+    UserTaskInfo userTaskInfo = getUserTaskByUserTaskId(userTaskId, httpServletRequest);
 
-    if (operationFutures != null) {
+    if (!asyncComplete && userTaskInfo != null) {
+      throw new IllegalArgumentException(
+          String.format("Sync operation should not have %s", USER_TASK_HEADER_NAME));
```

Comment:
Is there a particular reason that sync request should not have UUID?
User can still use UUID to retrieve the previous result, this might be useful for endpoint like `kafka_cluster_state`.

Reply:
It might be more appropriate to rewrite `/kafka_cluster_state` as async request? Since `kafka_cluster_state` takes longer than rest of sync requests. I consider each sync request as one-off and one should not need a UUID to retrieve the previous result.

Reply:
I agree with @kidkun on this one. Users should be able to retrieve the response of sync requests using UUIDs. 

Actually, I am confused and might be missing a basic point -- how would users retrieve the response of sync requests with the current implementation?